File: scratch/find_visited_wa_links.ts
import { config } from "dotenv";
import { cert, getApps, initializeApp } from "firebase-admin/app";
import { getFirestore } from "firebase-admin/firestore";
import * as fs from "fs";
import * as path from "path";
import * as sqlite3 from "sqlite3";

config({ path: ".env.local" });

const HISTORY_PATHS = [
  "/Users/tirthpatel/Library/Application Support/Google/Chrome/Default/History",
  "/Users/tirthpatel/Library/Application Support/BraveSoftware/Brave-Browser/Default/History",
  "/Users/tirthpatel/Library/Application Support/Arc/User Data/Default/History",
  "/Users/tirthpatel/Library/Application Support/Microsoft Edge/Default/History",
  "/Users/tirthpatel/Library/Application Support/com.operasoftware.Opera/History"
];

async function main() {
  const projectId = process.env.FIREBASE_PROJECT_ID;
  const clientEmail = process.env.FIREBASE_CLIENT_EMAIL;
  const privateKey = process.env.FIREBASE_PRIVATE_KEY?.replace(/\\n/g, "\n");

  if (!projectId || !clientEmail || !privateKey) {
    throw new Error("Missing Firebase config in .env.local");
  }

  if (!getApps().length) {
    initializeApp({ credential: cert({ projectId, clientEmail, privateKey }) });
  }

  const db = getFirestore();
  const mishwaUid = "A1gwf0ASuBOsbP1kgVDPVlQfbRw1";

  // 1. Fetch Mishwa's assigned contacts
  const contactsSnap = await db.collection("contacts").where("assignedTo", "==", mishwaUid).get();
  console.log(`Fetched ${contactsSnap.size} contacts assigned to Mishwa.`);

  const mishwaContacts = contactsSnap.docs.map(doc => {
    const d = doc.data();
    const rawMobile = (d.mobile || "").replace(/\D/g, "");
    const last10Digits = rawMobile.slice(-10);
    return {
      id: doc.id,
      name: d.name,
      email: d.email,
      mobile: d.mobile,
      rawMobile,
      last10Digits,
    };
  });

  // 2. Helper to query a single SQLite history DB
  function queryHistoryDB(dbPath: string): Promise<string[]> {
    return new Promise((resolve) => {
      if (!fs.existsSync(dbPath)) {
        resolve([]);
        return;
      }
      
      // Copy history file to a temp file because browser may lock it
      const tempPath = `${dbPath}.temp`;
      try {
        fs.copyFileSync(dbPath, tempPath);
      } catch (err) {
        resolve([]);
        return;
      }

      const sqlDb = new sqlite3.default.Database(tempPath, sqlite3.default.OPEN_READONLY, (err) => {
        if (err) {
          try { fs.unlinkSync(tempPath); } catch {}
          resolve([]);
          return;
        }
      });

      const urls: string[] = [];
      const sql = "SELECT url FROM urls WHERE url LIKE '%wa.me%' OR url LIKE '%whatsapp.com/send%'";
      sqlDb.all(sql, [], (err, rows: any[]) => {
        if (!err && rows) {
          rows.forEach(r => urls.push(r.url));
        }
        sqlDb.close(() => {
          try { fs.unlinkSync(tempPath); } catch {}
          resolve(urls);
        });
      });
    });
  }

  // 3. Search browser history files
  const visitedUrls = new Set<string>();
  for (const hPath of HISTORY_PATHS) {
    const urls = await queryHistoryDB(hPath);
    if (urls.length > 0) {
      console.log(`Found ${urls.length} WhatsApp links in: ${hPath}`);
      urls.forEach(u => visitedUrls.add(u));
    }
  }

  console.log(`\nUnique visited WhatsApp URLs: ${visitedUrls.size}`);

  // 4. Match phone numbers
  const matchedContacts: any[] = [];
  visitedUrls.forEach(url => {
    // Extract numbers from url
    const numbersInUrl = url.replace(/\D/g, "");
    mishwaContacts.forEach(c => {
      if (c.last10Digits && c.last10Digits.length >= 10 && numbersInUrl.includes(c.last10Digits)) {
        matchedContacts.push(c);
      }
    });
  });

  // Deduplicate matched contacts
  const uniqueMatches = Array.from(new Map(matchedContacts.map(c => [c.id, c])).values());

  console.log(`\n=== Matches with Mishwa's Assigned Tasks: ${uniqueMatches.length} ===`);
  uniqueMatches.forEach((c, idx) => {
    console.log(`${idx + 1}: ID: ${c.id} | Name: ${c.name} | Mobile: ${c.mobile} | Email: ${c.email}`);
  });
}

main().catch(console.error);

File: scratch/find_visited_wa_links.py
import os
import sqlite3
import shutil
import re
import json

HISTORY_PATHS = [
    "/Users/tirthpatel/Library/Application Support/Google/Chrome/Default/History",
    "/Users/tirthpatel/Library/Application Support/BraveSoftware/Brave-Browser/Default/History",
    "/Users/tirthpatel/Library/Application Support/Arc/User Data/Default/History",
    "/Users/tirthpatel/Library/Application Support/Microsoft Edge/Default/History",
    "/Users/tirthpatel/Library/Application Support/com.operasoftware.Opera/History"
]

def query_history(db_path):
    if not os.path.exists(db_path):
        return []
    
    temp_path = db_path + ".temp"
    try:
        shutil.copyfile(db_path, temp_path)
    except Exception as e:
        print(f"Error copying {db_path}: {e}")
        return []
        
    urls = []
    try:
        conn = sqlite3.connect(temp_path)
        cursor = conn.cursor()
        cursor.execute("SELECT url FROM urls WHERE url LIKE '%wa.me%' OR url LIKE '%whatsapp.com/send%'")
        rows = cursor.fetchall()
        urls = [r[0] for r in rows]
        conn.close()
    except Exception as e:
        print(f"Error reading SQLite db {db_path}: {e}")
    finally:
        if os.path.exists(temp_path):
            os.remove(temp_path)
            
    return urls

def main():
    print("Searching browser histories for WhatsApp links...")
    all_urls = set()
    for path in HISTORY_PATHS:
        urls = query_history(path)
        if urls:
            print(f"Found {len(urls)} WhatsApp links in: {path}")
            for u in urls:
                all_urls.add(u)
                
    print(f"Total unique visited WhatsApp URLs: {len(all_urls)}")
    
    # Extract numbers
    numbers = set()
    for url in all_urls:
        # Match digits in url after wa.me/ or inside send?phone=
        # E.g. https://wa.me/919974202309?text=...
        # E.g. https://api.whatsapp.com/send/?phone=919358518089&text=...
        matches = re.findall(r'\d+', url)
        for m in matches:
            if len(m) >= 10:  # Valid phone numbers are at least 10 digits
                numbers.add(m)
                
    numbers_list = list(numbers)
    print(f"Extracted {len(numbers_list)} unique phone numbers (>= 10 digits).")
    
    output_path = "/Users/tirthpatel/Downloads/team-app-main/scratch/visited_wa_numbers.json"
    with open(output_path, "w") as f:
        json.dump(numbers_list, f, indent=2)
    print(f"Saved phone numbers to {output_path}")

if __name__ == "__main__":
    main()
